h2/stream.py

```diff
@@ -269,7 +289,23 @@ def recv_push_promise(self, previous_state):
         event.parent_stream_id = self.stream_id
         return [event]
 
-    def send_reset(self, previous_state):
+    def send_end_stream(self, previous_state):
+        """
+        Called when an attempt is made to send END_STREAM at HALF_CLOSED_REMOTE
+        state.
+        """
+        assert previous_state == StreamState.HALF_CLOSED_REMOTE
+        self.stream_closed_by = StreamClosedBy.SEND_END_STREAM
+
+    def send_reset_stream(self, previous_state):
+        """
+        Called when an attempt is made to send RST_STREAM at non-closed stream
+        state.
+        """
+        assert previous_state != StreamState.CLOSED
+        self.stream_closed_by = StreamClosedBy.SEND_RST_STREAM
+
+    def reset_stream(self, previous_state):
```

Comment:
I think this needs a better name, as it's now misleadingly close to `send_reset_stream`. Something that is close to what it's actually for, which is maybe more like "invalid_frame_received".